src/Core/Geometry/DownSample.cpp

```diff
@@ -103,6 +104,74 @@ std::shared_ptr<PointCloud> SelectDownSample(const PointCloud &input,
+        const std::vector<size_t> &indices)
+{
+    auto output = std::make_shared<TriangleMesh>();
+    bool has_normals = input.HasVertexNormals();
+    bool has_colors = input.HasVertexColors();
+    // For each vertex, list face indices.
+    std::vector<std::vector<int>> vertex_to_triangle_temp(input.vertices_.size());
+    int triangle_id = 0;
+    for (auto trangle : input.triangles_) {
+        for (int i=0; i<3; i++)
+            vertex_to_triangle_temp[trangle(i)].push_back(triangle_id);
+        triangle_id++;
+    }
+    // Remove face indices of vertex_to_triangle_temp
+    // if it does not correspond to selected vertices
+    std::vector<std::vector<int>> vertex_to_triangle(input.vertices_.size());
+    for (auto vertex_id : indices) {
+        vertex_to_triangle[vertex_id] = vertex_to_triangle_temp[vertex_id];
+    }
+    // Make a triangle_to_vertex using vertex_to_triangle
+    std::vector<std::vector<int>> triangle_to_vertex(input.triangles_.size());
+    int vertex_id = 0;
+    for (auto face_ids : vertex_to_triangle) {
+        for (auto face_id : face_ids)
+            triangle_to_vertex[face_id].push_back(vertex_id);
+        vertex_id++;
+    }
+    // Only a face with three selected points contributes to mark mask_observed_vertex.
+    std::vector<bool> mask_observed_vertex(input.vertices_.size());
+    for (auto vertex_ids : triangle_to_vertex) {
+        if ((int)vertex_ids.size() == 3)
+            for (int i=0; i<3; i++)
+                mask_observed_vertex[vertex_ids[i]] = true;
+    }
+    // Rename vertex id based on selected points
+    std::vector<int> new_vertex_id(input.vertices_.size());
+    for (int i=0, cnt=0; i<mask_observed_vertex.size(); i++) {
+        if (mask_observed_vertex[i])
+            new_vertex_id[i] = cnt++;
+    }
+    // Push a triangle that has 3 selected vertices.
+    triangle_id = 0;
+    for (auto vertex_ids : triangle_to_vertex) {
+        if ((int)vertex_ids.size() == 3) {
+            Eigen::Vector3i new_face;
+            for (int i=0; i<3; i++)
+                new_face(i) = new_vertex_id[input.triangles_[triangle_id][i]];
+            output->triangles_.push_back(new_face);
+        }
+        triangle_id++;
+    }
+    // Push marked vertex.
+    for (int i=0; i<mask_observed_vertex.size(); i++) {
+        if (mask_observed_vertex[i]) {
+            output->vertices_.push_back(input.vertices_[i]);
+            if (has_normals) output->vertex_normals_.push_back(
+                    input.vertex_normals_[i]);
+            if (has_colors) output->vertex_colors_.push_back(
+                    input.vertex_colors_[i]);
+        }
+    }
+    PrintDebug("Triangle mesh sampled from %d vertices and %d triangles to %d vertices and %d triangles.\n",
+            (int)input.vertices_.size(), (int)input.triangles_.size(),
+            (int)output->vertices_.size(), (int)output->triangles_.size());
+    return output;
+}
```

Comment:
This function is bit long. I have tried to figure out the most efficient way to implement an algorithm that picks faces having three selected points. Maybe the names of variables are not very concise.

Reply:
I am not very sure this function can be placed here. In DownSample.cpp, most of the functions are point cloud operations.